test/examples/uct_hello_world.c

```diff
@@ -290,64 +290,87 @@ static ucs_status_t init_iface(char *dev_name, char *tl_name,
 static ucs_status_t dev_tl_lookup(const cmd_args_t *cmd_args,
                                   iface_info_t *iface_p)
 {
-    uct_md_resource_desc_t  *md_resources; /* Memory domain resource descriptor */
+    uct_component_h         *components;
+    unsigned                num_components;
+    unsigned                cmpt_index;
+    uct_component_attr_t    component_attr;
+    unsigned                md_index;
     uct_tl_resource_desc_t  *tl_resources; /* Communication resource descriptor */
-    unsigned                num_md_resources; /* Number of memory domains */
     unsigned                num_tl_resources; /* Number of transport resources resource objects created */
+    unsigned                tl_index;
     uct_md_config_t         *md_config;
     ucs_status_t            status;
-    int                     i;
-    int                     j;
 
-    status = uct_query_md_resources(&md_resources, &num_md_resources);
-    CHKERR_JUMP(UCS_OK != status, "query for memory domain resources", error_ret);
-
-    iface_p->iface = NULL;
-
-    /* Iterate through memory domain resources */
-    for (i = 0; i < num_md_resources; ++i) {
-        status = uct_md_config_read(md_resources[i].md_name, NULL, NULL, &md_config);
-        CHKERR_JUMP(UCS_OK != status, "read PD config", release_md);
-
-        status = uct_md_open(md_resources[i].md_name, md_config, &iface_p->md);
-        uct_config_release(md_config);
-        CHKERR_JUMP(UCS_OK != status, "open memory domains", release_md);
-
-        status = uct_md_query_tl_resources(iface_p->md, &tl_resources, &num_tl_resources);
-        CHKERR_JUMP(UCS_OK != status, "query transport resources", close_md);
-
-        /* Go through each available transport and find the proper name */
-        for (j = 0; j < num_tl_resources; ++j) {
-            if (!strcmp(cmd_args->dev_name, tl_resources[j].dev_name) &&
-                !strcmp(cmd_args->tl_name, tl_resources[j].tl_name)) {
-                status = init_iface(tl_resources[j].dev_name,
-                                    tl_resources[j].tl_name,
-                                    cmd_args->func_am_type, iface_p);
-                if (UCS_OK == status) {
-                    fprintf(stdout, "Using %s with %s.\n",
-                            tl_resources[j].dev_name,
-                            tl_resources[j].tl_name);
-                    fflush(stdout);
-                    uct_release_tl_resource_list(tl_resources);
-                    goto release_md;
+    status = uct_query_components(&components, &num_components);
+    CHKERR_JUMP(UCS_OK != status, "query for components", error_ret);
+
+    for (cmpt_index = 0; cmpt_index < num_components; ++cmpt_index) {
+
+        component_attr.field_mask = UCT_COMPONENT_ATTR_FIELD_MD_RESOURCE_COUNT;
+        status = uct_component_query(components[cmpt_index], &component_attr);
```

Comment:
minor: querying of components (number and themselves) is done several times thru the code. Separate func can be worth adding.

Reply:
i see it only 2 times in the code (line 310 and line 317), what am ia missing?

Reply:
It is relevant for UCP, libperf and tlinfo as well. I meant an auxiliary func, which would request memory domain resources (do component query twice + allocation).  Or can we unite UCT_COMPONENT_ATTR_FIELD_MD_RESOURCE_COUNT and UCT_COMPONENT_ATTR_FIELD_MD_RESOURCES in one call, so that allocation is done in UCT?

Reply:
i don't think there should be such common aux func, because this is an example which should be self-contained. Also, i didn't want to have allocation in UCT, to avoid the need to have 'release' API in UCT, and to let user allocate memory in own preferred method (for example, on the stack, as part of std::vector, etc)

Reply:
ok